### PR TITLE
Simulator for PLL

### PR DESCRIPTION
**Pull request by sugamarora99**

### CODE REVIEW

File: BeVolt/BSP/Simulator/DataGeneration/PLL.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+"""
+@brief change the clock frequency
+    Function to be called by simulate.py
+"""
+CurrentFrequency = 16000000
+
+
+def Change_Frequency(newFrequency):
+    with open(file, 'w') as csvfile:
```
Comment (SijWoo) [inline BeVolt/BSP/Simulator/DataGeneration/PLL.py:13]: Add this line `os.makedirs(os.path.dirname(file), exist_ok=True)` before opening the file. This checks to see if the Data directory exists and if it doesn't, then it'll create it to place the PLL.csv file.

File: BeVolt/BSP/Simulator/DataGeneration/PLL.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import csv
+
```
Comment (SijWoo) [inline BeVolt/BSP/Simulator/DataGeneration/PLL.py:2]: Add `import os` to execute `os.makedirs(os.path.dirname(file), exist_ok=True)`

**Comment by SijWoo** [comment]
I'm getting an error saying `name Current_Frequency is not defined` when I start simulator.py. I'm getting a segfault when I run the C code but that's because of Voltage_Init not having the necessary BSP_SPI code merged. Is there a Test_BSP_PLL file that you created Sugam?

**Comment by sugamarora99** [comment]
> I'm getting an error saying `name Current_Frequency is not defined` when I start simulator.py. I'm getting a segfault when I run the C code but that's because of Voltage_Init not having the necessary BSP_SPI code merged. Is there a Test_BSP_PLL file that you created Sugam?

Yeah

File: BeVolt/BSP/Simulator/DataGeneration/simulate.py [hunk 3]
```diff
@@ -53,16 +62,21 @@ def configure():
     # Get configuration settings
     global state, mode
     print("Welcome to the BPS Simulator")
-    print("Would you like to simulate 'charging' or 'discharging'?")
+    print("Would you like to simulate 'charging', 'discharging', or 'PLL'?")
     state = input()
-    while state != 'charging' and state != 'discharging':
-        print("That is not a valid option. Please enter 'charging' or 'discharging': ")
+    while state != 'charging' and state != 'discharging' and state != 'PLL':
+        print("That is not a valid option. Please enter 'charging', 'discharging', or 'PLL': ")
```
Comment (SijWoo) [inline BeVolt/BSP/Simulator/DataGeneration/simulate.py:68]: I saw on the CAN simulator pull request Rishi mentioned adding the any configuration stuff for whenever the user pressed CTRL+C. Specifically, handling it in the configure function

File: BeVolt/BSP/Simulator/DataGeneration/PLL.py (hunk 1 above)
Comment (SijWoo) [reply BeVolt/BSP/Simulator/DataGeneration/PLL.py:13]: Also add locks whenever accessing the csv file. Under issues, there is a ticket called _Simulator: Locking files during read and writes_ which shows what to add to lock the code

**Comment by chaseblock** [comment]
What's the status on this? The simulate.py does not seem to work for me. Just getting `list index out of range` after typing in `start`.

Is anyone else experiencing this issue?

**Comment by sugamarora99** [comment]
I am not getting this error

**Comment by sugamarora99** [comment]
> What's the status on this? The simulate.py does not seem to work for me. Just getting `list index out of range` after typing in `start`.
> 
> Is anyone else experiencing this issue?

I'm not getting this error

File: BeVolt/BSP/Simulator/Src/BSP_PLL.c [hunk 4]
```diff
@@ -1,12 +1,28 @@
...
+#include "simulator_conf.h"
+
+static char* file = GET_CSV_PATH(PLL_CSV_FILE);
 
 /**
  * @brief   Initialize the PLL so the system core frequency runs at your preferred frequency.
  * @param   None
  * @return  None
  */
 void BSP_PLL_Init(void) {
```
Comment (ClarkPoon) [inline BeVolt/BSP/Simulator/Src/BSP_PLL.c:14]: This function should set the clock frequency to 80MHz. On the actual board, the clock is set to 16MHz automatically on startup, and this function is used to change the clock frequency to 80MHz

**Pull request merged**